bestla/bestla/bestla_parallel.h

```diff
@@ -23,10 +24,238 @@
+        time_per_p = (time_p - sync_time[0].load()) / (1.0 * (mThreadNum - cr->E_core_num));
+        time_per_e = (time_e - sync_time[1].load()) / (1.0 * cr->E_core_num);
+      } else {
+        for (size_t i = 0; i < mThreadNum - 1; i++) {
+          func_[i] = &func;
+        }
+        func(0);
+        while (true) {
+          if (running.load() == 0)
+            break;
+          else
+            _mm_pause();
+        }
+      }
+    } else {
+      func(0);
+    }
+  }
+
+  void set_threads(int nthreads) override {
+    if (nthreads != mThreadNum) {
+      stop_threads();
+      mThreadNum = nthreads;
+      cr = &device::CpuRuntime::getInstance(nthreads);
+      create_threads();
+    }
+  }
+
+  inline void sync(int idx = 0) override {
+    flag[idx].fetch_sub(1);
+    if (cr->mHybrid) {
+      utils::timer tm;
+      tm.start();
+      while (true) {
+        if (flag[idx].load() == 0)
+          break;
+        else
+          _mm_pause();
+      }
+      sync_time[device::CpuDevice::isEcore()].fetch_add(int(tm.stop()));
+    } else {
+      while (true) {
+        if (flag[idx].load() == 0)
+          break;
+        else
+          _mm_pause();
+      }
+    }
+  }
+
+  std::pair<float, float> get_PEtime() const override { return {time_per_p, time_per_e}; };
+
+  ~StdThreading() { stop_threads(); }
+
+ private:
+  void stop_threads() {
+    stop = true;
+    for (int i = 0; i < mThreadNum - 1; i++) thdset[i].join();
+    thdset.clear();
+    // printf("stop %d\n", mThreadNum);
+  }
+  void create_threads() {
+    // printf("create %d\n", mThreadNum);
+    thdset.resize(mThreadNum - 1);
+    stop = false;
+    GetCPUDevice();
+    int* core_order;  // Note: client CPU earlier than Gen 12th will get poor perf when not using all cores.
+    if (_cd->isHybrid()) {
+      core_order = new int[_cd->getThreads()];
+      memcpy(reinterpret_cast<void*>(core_order), reinterpret_cast<void*>(_cd->getPCores()),
+             _cd->getPcoreNum() * sizeof(int));
+      memcpy(reinterpret_cast<void*>(core_order + _cd->getPcoreNum()), reinterpret_cast<void*>(_cd->getECores()),
+             _cd->getEcoreNum() * sizeof(int));
+      memcpy(reinterpret_cast<void*>(core_order + _cd->getPcoreNum() + _cd->getEcoreNum()),
+             reinterpret_cast<void*>(_cd->getSMTCores()), _cd->getSMTcoreNum() * sizeof(int));
+    } else {
+      core_order = new int[mThreadNum];
+      for (int i = 0; i < mThreadNum; i++) core_order[i] = i;
+    }
+    _cd->core_bond(core_order[0]);
+    if (cr->mHybrid) {
+      thread_time.resize(mThreadNum - 1);
+      for (size_t i = 0; i < mThreadNum - 1; i++) {
+        thdset[i] = std::thread(
+            [&](int tidx, int core_id) {
+              _cd->core_bond(core_id);
+              utils::timer tm;
+              while (true) {
+                if (stop.load() == true) break;
+                if (func_[tidx] != nullptr) {
+                  tm.start();
+                  (*func_[tidx])(tidx + 1);
```

Comment:
why we call func in create_thread?

Reply:
The code of new thread has been fixed when thread created. We leave some interface (here is func_) to run different code assigned to thread in using.